File: services/tu-core/internal/domain/academic/models/valueobjects/fullname.go
package valueobjects

import (
	"errors"
	"strings"
)

// FullName adalah value object untuk nama lengkap.
type FullName struct {
	firstName  string
	middleName string
	lastName   string
}

// NewFullName membuat FullName baru dengan validasi.
func NewFullName(firstName, middleName, lastName string) (FullName, error) {
	if strings.TrimSpace(firstName) == "" {
		return FullName{}, errors.New("nama depan tidak boleh kosong")
	}
	return FullName{
		firstName:  strings.TrimSpace(firstName),
		middleName: strings.TrimSpace(middleName),
		lastName:   strings.TrimSpace(lastName),
	}, nil
}

// Getters

func (n FullName) FirstName() string  { return n.firstName }
func (n FullName) MiddleName() string { return n.middleName }
func (n FullName) LastName() string   { return n.lastName }

// DisplayName mengembalikan nama lengkap untuk ditampilkan.
func (n FullName) DisplayName() string {
	parts := []string{n.firstName}
	if n.middleName != "" {
		parts = append(parts, n.middleName)
	}
	if n.lastName != "" {
		parts = append(parts, n.lastName)
	}
	return strings.Join(parts, " ")
}

// Equals mengecek kesamaan.
func (n FullName) Equals(other FullName) bool {
	return n.firstName == other.firstName &&
		n.middleName == other.middleName &&
		n.lastName == other.lastName
}

File: apps/siakad-tu/src/domain/student/value-objects/FullName.js
// apps/siakad-tu/src/domain/student/value-objects/FullName.js

/**
 * Value Object: FullName
 *
 * Terdiri dari firstName, middleName (optional), lastName (optional)
 * Aturan: firstName wajib, lainnya optional
 */
export class FullName {
  #firstName
  #middleName
  #lastName

  constructor(firstName, middleName = '', lastName = '') {
    if (!firstName || String(firstName).trim() === '') {
      throw new Error('Nama depan (firstName) tidak boleh kosong.')
    }

    this.#firstName = String(firstName).trim()
    this.#middleName = middleName ? String(middleName).trim() : ''
    this.#lastName = lastName ? String(lastName).trim() : ''

    Object.freeze(this)
  }

  get firstName() {
    return this.#firstName
  }
  get middleName() {
    return this.#middleName
  }
  get lastName() {
    return this.#lastName
  }

  /**
   * Nama lengkap untuk ditampilkan
   */
  get displayName() {
    return [this.#firstName, this.#middleName, this.#lastName].filter(Boolean).join(' ')
  }

  equals(other) {
    if (!(other instanceof FullName)) return false
    return (
      this.#firstName === other.#firstName &&
      this.#middleName === other.#middleName &&
      this.#lastName === other.#lastName
    )
  }

  toJSON() {
    return {
      firstName: this.#firstName,
      middleName: this.#middleName,
      lastName: this.#lastName,
    }
  }

  static fromJSON(data) {
    return new FullName(data.firstName, data.middleName, data.lastName)
  }
}
